core/storage/pager.rs

```diff
@@ -587,15 +594,60 @@ impl Pager {
                         BtreePageAllocMode::Exact(root_page_num),
                     )?;
                     let allocated_page_id = page.get().get().id as u32;
+                    drop(page);
+
+                    //  We want to be allocated root_page_num but we got allocated allocated_page_id. We need to swap these two pages so that
+                    //  root pages remain clustered at the beginning of the file
                     if allocated_page_id != root_page_num {
-                        //  TODO(Zaid): Handle swapping the allocated page with the desired root page
-                    }
+                        //  save all the btree cursors so that all page references are freed
+                        for cursor in cursors {
+                            if let Some(cursor) = cursor {
+                                let Cursor::BTree(c) = cursor else {
+                                    continue;
+                                };
+                                c.save_context_external_invalidation()?;
```

Comment:
what happens if this returns `IO`? do we allocate a page again on line 291 when we re-enter this function?

Reply:
@jussisaurio yeah, this requires a state machine to be implemented for autovacuum. Pere brought this up in an [earlier comment](https://github.com/tursodatabase/turso/pull/1874#discussion_r2202380974) as well.

My plan is to port over the remaining autovacuum logic and then define the state machine at the end because we'll have all the invariants defined at that point.